### ISSUE TITLE
Having Trouble Recreating Results in the Paper

### ISSUE THREAD

**Issue by yulun-rayn** [Sachithx/EntroPE#6]
Hi,

Thanks for releasing the source code! I did a quick run on ETTh1 using the exact configs in [here](https://github.com/Sachithx/EntroPE/blob/main/scripts/etth1.sh) and did not manage to create the results in the [paper](https://arxiv.org/pdf/2509.26157). I got 0.438 MSE and 0.432 MAE. Run met early stopping criteria on epoch 40. Before I proceed to run more experiments, could you clarify if the configs in here are the same as what were used for the paper? Am I missing something? Thanks a lot!

**Comment by Sachithx** [comment]
Thank you, @yulun-rayn, for your interest in our work! 🙏
The issue you mentioned has already been discussed and resolved in [this thread](https://github.com/Sachithx/EntroPE/issues/2). Please have a look there for the detailed explanation.
If you continue to encounter similar issues or have any other questions, feel free to open a new issue. I’d be happy to look into it and help further.

**Comment by Sachithx** [comment]
Below is the resulting outcome based on the scripts in the repository.
<img width="1357" height="783" alt="Image" src="https://github.com/user-attachments/assets/596c4847-ec07-4660-b2a6-fe96a9775eaf" />

<img width="2528" height="1328" alt="Image" src="https://github.com/user-attachments/assets/03bba724-87c4-434c-998e-afc91ee3b66f" />

**Comment by yulun-rayn** [comment]
I don't think how the threshold determines patch size would be affected by any dependencies or GPUs? The current etth1 entropy checkpoint & settings just give me all false `patch_start_mask` and then `max_patch_length` makes all patch size 12.

**Comment by yulun-rayn** [comment]
Is it possible that the entropy model you're using is not the same as the one currently resides in the repo?

**Comment by yulun-rayn** [comment]
and the datasets you used are the exact same ones as these? https://github.com/zhouhaoyi/ETDataset/tree/main/ETT-small

**Comment by Sachithx** [comment]
> and the datasets you used are the exact same ones as these? https://github.com/zhouhaoyi/ETDataset/tree/main/ETT-small

yes.
Can you check with 5 different seeds? @yulun-rayn

**Comment by Sachithx** [comment]
> Ok I will. Thanks for your help! I just tested on another device. Just to make sure I'm looking at the right number, are the MSE, MAE, RSE printouts after early stopping not the ones to look at? I guess if you look at lowest test loss among all epochs, you can get as low as 0.430, which is better than the printouts even though not at the level of those reported in paper. Also, if we record best test loss in any epoch, how do we get the corresponding MAE?
> 
> ```
> Epoch: 40 cost time: 15.153807640075684
> Epoch: 40, Steps: 65 | Train Loss: 0.4155461 Vali Loss: 1.0248138 Test Loss: 0.4295133
> EarlyStopping counter: 20 out of 20
> Early stopping
> >>>>>>>testing: EntroPE_ETTh1_SL96_PL192_GL1_EL1_DL1_GD16_DD16_ED16_GH4_EH4_DH4<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
> test 2689
> MSE: 0.43724003434181213, MAE: 0.4320778250694275, RSE: 0.6279306411743164
> wandb: 
> wandb: Run history:
> wandb:            batch_loss ▇█▇▇▅▅▄▄▃▂▂▃▂▃▂▂▂▁▃▁▂▂▂▂▂▂▂▂▂▁▂▂▂▂▂▁▁▂▁
> wandb:         best_val_loss ▁
> wandb:  early_stopping_epoch ▁
> wandb:                 epoch ▁▁▁▂▂▂▂▂▂▃▃▃▃▃▄▄▄▄▄▄▅▅▅▅▅▅▆▆▆▆▆▇▇▇▇▇▇███
> wandb:        epoch_duration █▃▂▂▂▂▂▂▂▂▂▁▂▂▁▁▂▂▂▂▃▂▂▂▃▂▂▂▂▂▂▂▂▄▃▅▄▄▄▄
> wandb:       epoch_test_loss █▇▆▅▄▃▃▂▂▂▂▂▂▁▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
> wandb:      epoch_train_loss ██▇▇▆▅▄▃▃▃▂▂▂▂▂▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
> wandb:        epoch_val_loss █▇▆▅▄▃▃▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
> wandb: estimated_time_left_s █▇▇▇▆▆▆▆▆▆▅▅▅▅▄▄▄▄▄▄▄▃▃▃▃▂▂▂▂▂▂▃▂▁▁▂▁▁▁
> wandb:         learning_rate ▁▁▁▁▁▁▂▂▂▂▂▂▃▃▃▃▄▄▄▅▅▅▅▆▆▆▆▇▇▇▇▇▇██████
> wandb:                   +13 ...
> wandb: 
> wandb: Run summary:
> wandb:            batch_loss 0.41733
> wandb:         best_val_loss 1.00811
> wandb:  early_stopping_epoch 40
> wandb:                 epoch 40
> wandb:        epoch_duration 15.15381
> wandb:       epoch_test_loss 0.42951
> wandb:      epoch_train_loss 0.41555
> wandb:        epoch_val_loss 1.02481
> wandb: estimated_time_left_s 146115.71646
> wandb:         learning_rate 0.001
> wandb:                   +13 ...
> wandb: 
> ```

I am only recording the final saved model based on the validation loss, as a result of early stopping or the best model. Finally script auto-loads the saved model and tests again; those result is saved in the result.txt file (which is equal to the final line of your snap). I extract the result from there. Example as below:.

<img width="646" height="53" alt="Image" src="https://github.com/user-attachments/assets/9c255073-16f6-476a-b403-30e35fd4a87a" />

**Comment by yulun-rayn** [comment]
yeah I noticed that too. didn't think it'd make that much of a difference but I could be wrong. I'm just running interactively on a10g instance to test it out for now. what is your device?

**Comment by Sachithx** [comment]
> You're doing great! Thanks for the help and prompt responses. I'll test a few different seeds and see how it goes. No further question for now :)

I used a different GPU (NVIDIA GeForce RTX 4090 D), but there was still a slight deviation. It usually reports the best model around 0.422. However, with the new GPU, the validation loss didn’t decrease, which led me to skip saving and save at 0.425. 

<img width="556" height="396" alt="Image" src="https://github.com/user-attachments/assets/c04f9639-c341-48e9-b53d-473562e56078" />

<img width="454" height="361" alt="Image" src="https://github.com/user-attachments/assets/8bcd5963-98fd-495e-b3e3-4e3e32d6be1d" />

**Comment by yulun-rayn** [comment]
I ran 5 seeds on L40S. On PL192, MSEs are 0.435, 0.436, 0.427, 0.432, 0.431. Here are the final log for the last run

```
Epoch: 44, Steps: 65 | Train Loss: 0.4093412 Vali Loss: 1.0256162 Test Loss: 0.4326622
EarlyStopping counter: 19 out of 20
    iters: 1, epoch: 45 | loss: 0.4265893
    speed: 4.6114s/iter; left time: 16785.5529s
Epoch: 45 cost time: 2.145280122756958
Epoch: 45, Steps: 65 | Train Loss: 0.4086395 Vali Loss: 1.0140058 Test Loss: 0.4269505
EarlyStopping counter: 20 out of 20
Early stopping
>>>>>>>testing: EntroPE_ETTh1_SL96_PL192_GL1_EL1_DL1_GD16_DD16_ED16_GH4_EH4_DH4<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
test 2689
MSE: 0.4311450123786926, MAE: 0.4318341016769409, RSE: 0.623538613319397
wandb: updating run metadata
wandb: 
wandb: Run history:
wandb:            batch_loss ███▆▅▄▃▃▃▃▂▃▂▂▂▂▂▂▃▂▃▂▁▂▂▂▂▂▂▂▂▁▁▂▂▂▂▁▁▂
wandb:         best_val_loss ▁
wandb:  early_stopping_epoch ▁
wandb:                 epoch ▁▁▁▁▂▂▂▂▂▃▃▃▃▃▃▄▄▄▄▄▅▅▅▅▅▅▆▆▆▆▆▆▇▇▇▇▇███
wandb:        epoch_duration █▄▂▁▁▆▄▄▂▂▂▁▁▂▃▂▄▄▁▂▃▆▂▂▆▂▂▃▂▄▁▃▄▂▄▃▂▂▅▂
wandb:       epoch_test_loss █▇▆▅▄▄▃▂▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
wandb:      epoch_train_loss ██▇▇▆▅▄▃▃▂▂▂▂▂▂▂▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
wandb:        epoch_val_loss █▇▆▅▄▃▃▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
wandb: estimated_time_left_s █▇▇▇▅▇▆▆▅▇▄▄▅▄▄▄▄▅▄▄▃▅▄▃▃▃▄▃▃▃▂▂▂▃▁▂▂▂▁▁
wandb:         learning_rate ▁▁▁▁▁▁▂▂▂▂▂▃▃▃▃▄▄▄▄▅▅▆▆▆▆▇▇▇▇▇██████████
wandb:                   +13 ...
wandb: 
wandb: Run summary:
wandb:            batch_loss 0.42659
wandb:         best_val_loss 1.00844
wandb:  early_stopping_epoch 45
wandb:                 epoch 45
wandb:        epoch_duration 2.14528
wandb:       epoch_test_loss 0.42695
wandb:      epoch_train_loss 0.40864
wandb:        epoch_val_loss 1.01401
wandb: estimated_time_left_s 16785.55287
wandb:         learning_rate 0.00099
wandb:                   +13 ...
wandb: 
```

**Comment by Sachithx** [comment]
> Hello, thank you. I’ve checked this matter, and the configuration link here is no longer valid. I

Which was the fix that I made to main to add. missing layer_utils.py.
